### PR TITLE
Em/agl/kd/land framework

### PR DESCRIPTION
**Pull request by kmdeck**
# Description

This PR builds on the PR 1279, which added the src/Land/Model and test/Land/Model directories. PR 1279 created the LandModel framework (of which the SoilModel is a component, in turn composed of a water and a heat balance law). PR1279 also added SoilWaterParameterizations.jl, which contained the parameterized functions that we would need for the balance law for water in soil, and some unit tests for those functions (test_water_parameterizations.jl).  This PR adds in the balance law components and a boundary conditions interface for the soil water model. 

Our goal is to make it so that, within the soil model, one could run just the heat model, or just the water model, with different degrees of coupling (no coupling, or one way coupling in either direction). One way coupling is envisioned to occur when the user prescribes a function for e.g. water content as a function of space and time, instead of dynamically determining this from a PDE. This function could then be used to drive the heat model.

This PR adds in the balance law terms for the water component for two concrete types, inheriting from AbstractWaterModel. One is a PrescribedWaterModel, in which the user could supply a function of the water content as a function of space and time. There are no state variables or fluxes in this case. The other concrete type is SoilWaterModel, which has the necessary fluxes, etc, for Richard's equation. Along those lines, we created a function get_moisture_content(), which returns the state variables for moisture if a SoilWaterModel is used, or the value of the prescribed functions if a PrescribedWaterModel is chosen. When the balance law for heat is added, it will need to call that function, since moisture shows up in the equations for heat.

We also have a PrescribedTemperatureModel, along those same lines, and a get_temperature() function. Right now, all of the soil models use that AbstractHeatModel type. A second type, akin to the SoilWaterModel type, will be added in a future PR.

We've tested that (and have test cases showing that): 
(1) the two Prescribed models can be used together & the LandModel can still be integrated (nothing happens, no state variables, but the boundary condition calls dont give errors in this case). (prescribed_twice.jl)
(2) The way we supply boundary conditions (as a structure with Dirichlet or Neumann conditions, for the top or bottom of the domain) works as expected. (test_bc.jl)
(3) Richard's equation gives a solution which matches what we get from verified Matlab code (haverkamp_test.jl)
(4) in steady state, the gradient of the hydraulic head goes to zero. This is in tutorials, because it is more of a sanity check/benchmark, compared with a true unit or integration test. 

I have

- [X] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [X] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [X] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by kmdeck** [comment]
tests are failing after rebase initially because of EveryXSimulationTime changes. debugging...

**Comment by kmdeck** [comment]
After conversations with @charleskawczynski and @gagnelandmanna, some changes were made:
- renamed SoilParamSet to SoilParamFunctions to avoid confusion with param sets in general.
- made the default value for the volumetric ice fraction NaN rather than 0. I remembered that the reason I made this 0 was because then the default is Richard's eqn for liquid water. Ice only does stuff if freeze/thaw is present, but that is an additional source term the user will optionally add. It probably doesn't matter too much, though!
- defined soil_boundary_state and the various balance law methods for a general BalanceLaw argument. The prescribed models now default to those, so we only need one copy, rather than two. The more specific SoilWater (SoilHeat) model has (will have) its own method.
- moved SoilParamFunctions to be an attribute of the SoilModel, rather than the SoilWaterModel, since the heat model will eventually need those functions as well, and conceptually it makes sense for them to be a feature of Soil, rather than subparts of Soil.


Not done in this PR but discussed:
- consider making a wrapper for functions like pressure head and hydraulic conductivity
- we may not need to export vars_state functions in LandModel
- make orientation an attribute of the LandModel. Use init_aux and pass in orientation structure to that to generate auxiliary variables.

**Comment by kmdeck** [comment]
> After conversations with @charleskawczynski and @gagnelandmanna, some changes were made:
> 
>     * renamed SoilParamSet to SoilParamFunctions to avoid confusion with param sets in general.
> 
>     * made the default value for the volumetric ice fraction NaN rather than 0. I remembered that the reason I made this 0 was because then the default is Richard's eqn for liquid water. Ice only does stuff if freeze/thaw is present, but that is an additional source term the user will optionally add. It probably doesn't matter too much, though!
> 
>     * defined soil_boundary_state and the various balance law methods for a general BalanceLaw argument. The prescribed models now default to those, so we only need one copy, rather than two. The more specific SoilWater (SoilHeat) model has (will have) its own method.
> 
>     * moved SoilParamFunctions to be an attribute of the SoilModel, rather than the SoilWaterModel, since the heat model will eventually need those functions as well, and conceptually it makes sense for them to be a feature of Soil, rather than subparts of Soil.
> 
> 
> Not done in this PR but discussed:
> 
>     * consider making a wrapper for functions like pressure head and hydraulic conductivity
> 
>     * we may not need to export vars_state functions in LandModel
> 
>     * make orientation an attribute of the LandModel. Use init_aux and pass in orientation structure to that to generate auxiliary variables.

edit - theta_ice needs to be zero for Richard's equation alone, without freeze/thaw, because we dont have any constraint that says theta_ice+theta_water <= porosity. It doesnt make sense for theta_ice to be nonzero without freeze/thaw, I think.

changing this back, and making a small change to the docs, then will rebase.

File: test/Land/Model/haverkamp_test.jl [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+soil_param_functions = SoilParamFunctions(
+    porosity = 0.495,
+    Ksat = 0.0443 / (3600 * 100),
+    S_s = 1e-3,
+)
+# Keep in mind that what is passed is aux⁻
+# Fluxes are multiplied by ẑ (normal to the surface, -normal to the bottom,
+# where normal point outs of the domain.)
+surface_state = (aux, t) -> FT(0.494)
```
Comment (charleskawczynski) [inline test/Land/Model/haverkamp_test.jl:60]: Let's put this inside a @testset block

File: test/Land/Model/prescribed_twice.jl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+N_poly = 5;
+nelem_vert = 10;
+
+# Specify the domain boundaries
+zmax = FT(0);
+zmin = FT(-1)
+
+driver_config = ClimateMachine.SingleStackConfiguration(
+    "LandModel",
```
Comment (charleskawczynski) [inline test/Land/Model/prescribed_twice.jl:53]: Let's put this inside a @testset block

File: test/Land/Model/test_bc.jl [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+using ClimateMachine.DGMethods
+using ClimateMachine.DGMethods.NumericalFluxes
+using ClimateMachine.DGMethods: BalanceLaw, LocalGeometry
+using ClimateMachine.MPIStateArrays
+using ClimateMachine.GenericCallbacks
+using ClimateMachine.ODESolvers
+using ClimateMachine.VariableTemplates
+using ClimateMachine.SingleStackUtils
+
+FT = Float64;
```
Comment (charleskawczynski) [inline test/Land/Model/test_bc.jl:25]: Same, let's put this inside a @testset block

Comment (kmdeck) [reply test/Land/Model/test_bc.jl:25]: oh yeah! forgot that one!

File: test/Land/Model/haverkamp_test.jl [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+    aux_vars = SingleStackUtils.get_vars_from_nodal_stack(
+        mygrid,
+        aux,
+        vars_state_auxiliary(m, FT),
+    )
+    all_vars = OrderedDict(state_vars..., aux_vars...)
+    all_vars["t"] = [t]
+
+    # Compare with Bonan simulation data at 1 day.
+    bonan_moisture = reverse([
```
Comment (charleskawczynski) [inline test/Land/Model/haverkamp_test.jl:135]: Just to keep track, we should probably track this with Artifacts, as is done in Thermodynamics [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/test/Common/Thermodynamics/data_tests.jl).

Comment (kmdeck) [reply test/Land/Model/haverkamp_test.jl:133]: ok! I will add this to the list, now there are three items:
consider making a wrapper for functions like pressure head and hydraulic conductivity
add Bonan data using Artifacts
make orientation an attribute of the LandModel. Use init_aux and pass in orientation structure to that to generate auxiliary variables.

File: tutorials/Land/Soil/sanity_checks/equilibrium_test.jl [hunk 5]
```diff
@@ -0,0 +1,223 @@
+# This is not a unit test, but rather, a benchmark case/sanity check.
+using MPI
+using OrderedCollections
+using StaticArrays
+using Statistics
+using Plots
```
[the remark this answers is not in the document]
Comment (charleskawczynski) [reply tutorials/Land/Soil/sanity_checks/equilibrium_test.jl:6]: The tutorials are run in the `docs/` environment, which does have Plots. That said, it doesn't look like this tutorial is included in the [tutorials array](https://github.com/CliMA/ClimateMachine.jl/blob/master/docs/list_of_tutorials.jl#L21-L47).

[the remark this answers is not in the document]
Comment (kmdeck) [reply tutorials/Land/Soil/sanity_checks/equilibrium_test.jl:6]: ok! Ill try commenting it more clearly and adding it to that.

[the remark this answers is not in the document]
Comment (kmdeck) [reply tutorials/Land/Soil/sanity_checks/equilibrium_test.jl]: mm yes, had some issues (clima formater failed when I added in this to the list_of_tutorials.jl, so im making an error somewhere). Added this as a ticket on our project board!

**Comment by kmdeck** [comment]
@charleskawczynski maybe??

**Comment by kmdeck** [comment]
> Build failed (retrying...):
> 
>     * [test-os (ubuntu-latest)](https://github.com/CliMA/ClimateMachine.jl/runs/903639901)

@kpamnany would you mind trying this again?

**Comment by kpamnany** [comment]
It's going to retry automatically (that's what the message indicates). I will keep an eye on it though. :)

**Comment by kmdeck** [comment]
> Build failed:
> 
>     * test-os (windows-latest)

hmm, it failed again. (last time it failed on ubuntu...?). @kpamnany is there any error log this time?

**Comment by kpamnany** [comment]
[Here](https://github.com/CliMA/ClimateMachine.jl/runs/903775581?check_suite_focus=true#step:8:3118). A very mystifying message. (If you click the little red X next to the bors message about Merge 1342, you'll see the links, but it should be linked properly in the message... things are behaving very strangely today.)

File: slurmci-test.toml [hunk 6]
```diff
@@ -47,6 +47,9 @@ cpu_gpu = [
   { file = "test/Atmos/Parameterizations/Microphysics/KM_saturation_adjustment.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_warm_rain.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_ice.jl", slurmargs = ["--ntasks=3"], args = [] },
+  { file = "test/Land/Model/haverkamp_test.jl", slurmargs = ["--ntasks=1'"], args = [] },
```
Comment (kpamnany) [inline slurmci-test.toml:50]:
```suggestion
  { file = "test/Land/Model/haverkamp_test.jl", slurmargs = ["--ntasks=1"], args = [] },
```
This is breaking slurmci. :(

Comment (kmdeck) [reply slurmci-test.toml:50]: oh nooo i had an extra apostrophe! and is it something to do with the spaces too?

Comment (kmdeck) [reply slurmci-test.toml]: thank you so much for catching this!

File: slurmci-test.toml [hunk 7]
```diff
@@ -47,6 +47,9 @@ cpu_gpu = [
   { file = "test/Atmos/Parameterizations/Microphysics/KM_saturation_adjustment.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_warm_rain.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_ice.jl", slurmargs = ["--ntasks=3"], args = [] },
+  { file = "test/Land/Model/haverkamp_test.jl", slurmargs = ["--ntasks=1"], args = [] },
+  { file = "test/Land/Model/test_bc.jl", slurmargs = ["--ntasks=1'"], args = [] },
```
Comment (kpamnany) [inline slurmci-test.toml:51]:
```suggestion
  { file = "test/Land/Model/test_bc.jl", slurmargs = ["--ntasks=1"], args = [] },
```
Sorry, this one too.

File: slurmci-test.toml (hunk 6 above)
Comment (charleskawczynski) [reply slurmci-test.toml]: > oh nooo i had an extra apostrophe! and is it something to do with the spaces too?

Quite possibly, I'd try to stick very close to the existing pattern

File: slurmci-test.toml (hunk 7 above)
Comment (jakebolewski) [reply slurmci-test.toml:51]: maybe we need to add a lint check to validate toml files

Comment (jakebolewski) [reply slurmci-test.toml:51]: oh i see, it was just a slurm args issue

**Comment by kmdeck** [comment]
macro testsets is not defined on cluster somehow, will look into

**Comment by kmdeck** [comment]
> macro testsets is not defined on cluster somehow, will look into

needed to add "using Tests" since we are running on cluster, instead of in test suite?

**Comment by jakebolewski** [comment]
if you are running the file in it's own session / context then it needs it will need to import it's dependencies

File: slurmci-test.toml [hunk 8]
```diff
@@ -47,6 +47,9 @@ cpu_gpu = [
   { file = "test/Atmos/Parameterizations/Microphysics/KM_saturation_adjustment.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_warm_rain.jl", slurmargs = ["--ntasks=3"], args = [] },
   { file = "test/Atmos/Parameterizations/Microphysics/KM_ice.jl", slurmargs = ["--ntasks=3"], args = [] },
+  { file = "test/Land/Model/haverkamp_test.jl", slurmargs = ["--ntasks=1"], args = [] },
+  { file = "test/Land/Model/test_bc.jl", slurmargs = ["--ntasks=1"], args = [] },
+
```
Comment (charleskawczynski) [inline slurmci-test.toml:52]: Let's remove this empty line just in case it's problematic.

File: src/Land/Model/SoilWaterParameterizations.jl [hunk 9]
```diff
@@ -421,29 +424,44 @@ pressure head ψ; meters.
...
 case with an error.
-
 """
-function effective_saturation(porosity::FT, θ_l::FT) where {FT}
+function effective_saturation(porosity::FT, ϑ_l::FT) where {FT}
 
-    if θ_l < 0
-        throw(DomainError(θ_l, "Effective saturation is negative."))
+    if ϑ_l < 0
+        throw(DomainError(ϑ_l, "Effective saturation is negative."))
```
Comment (charleskawczynski) [inline src/Land/Model/SoilWaterParameterizations.jl:462]: This may not be GPU friendly. I'd just use `error` (see [this](https://github.com/CliMA/ClimateMachine.jl/blob/ck/update_heat_eq/src/Common/Thermodynamics/relations.jl#L1107-L1129) for GPU-friendly error handling).

Comment (kmdeck) [reply src/Land/Model/SoilWaterParameterizations.jl:462]: will take on next!

**Comment by jakebolewski** [comment]
Awesome work @kmdeck!  Sorry it was such a pain at the end.

**Comment by kmdeck** [comment]
yessssss i almost cant believe it! haha

@jakebolewski thank you for your help!! @kpamnany @charleskawczynski  also! :D
